pkg/network/server.go

```diff
@@ -813,11 +817,66 @@ func (s *Server) RelayTxn(t *transaction.Transaction) RelayReason {
 
 // broadcastTX broadcasts an inventory message about new transaction.
 func (s *Server) broadcastTX(t *transaction.Transaction) {
-	msg := s.MkMsg(CMDInv, payload.NewInventory(payload.TXType, []util.Uint256{t.Hash()}))
+	select {
+	case s.transactions <- t:
+	case <-s.quit:
+	}
+}
+
+func (s *Server) broadcastTxHashes(hs []util.Uint256) {
+	msg := s.MkMsg(CMDInv, payload.NewInventory(payload.TXType, hs))
 
 	// We need to filter out non-relaying nodes, so plain broadcast
 	// functions don't fit here.
 	s.iteratePeersWithSendMsg(msg, Peer.EnqueuePacket, func(p Peer) bool {
 		return p.Handshaked() && p.Version().Relay
 	})
 }
+
+// broadcastTxLoop is a loop for batching and sending
+// transactions hashes in and INV payload.
+func (s *Server) broadcastTxLoop() {
+	const (
+		batchTime = time.Millisecond * 50
+		batchSize = 32
+	)
+
+	txs := make([]util.Uint256, 0, batchSize)
+
+	timer := time.NewTimer(batchTime)
+	defer timer.Stop()
+
+	broadcast := func() {
+		go s.broadcastTxHashes(txs)
```

Comment:
That seems to be a little dangerous, have you tried not spawning a goroutine?

Reply:
Yes. The problem is slow peers can possibly hang all tx broadcasting.

Reply:
Yeah, but that's a known different problem, at the moment I fear more of unbalanced pipeline and goroutine leakage.